rfcs/0148-pipe-operator.md

```diff
@@ -0,0 +1,330 @@
+
+```nix
+defaultPrefsFile = pkgs.writeText "nixos-default-prefs.js" (
+  lib.concatStringsSep "\n" (
+    lib.mapAttrsToList
+    (
+      key: value: ''
+        // ${value.reason}
+        pref("${key}", ${builtins.toJSON value.value});
+      ''
+    )
+    defaultPrefs
+  )
+);
+```
+
+One can observe the following issues:
+
+- If you want to follow the data flow, you must read it from bottom to top,
+  from the inside to the outside (the input here is `defaultPrefs`).
+- Adding a function call to the output would require wrapping the entire
+  expression in parentheses and increasing its indentation.
+
+Compare this to the equivalent call with `lib.pipe`:
+
+```nix
+defaultPrefsFile = pipe defaultPrefs [
+  (lib.mapAttrsToList (
+    key: value: ''
+      // ${value.reason}
+      pref("${key}", ${builtins.toJSON value.value});
+    ''
+  ))
+  (lib.concatStringsSep "\n")
+  (pkgs.writeText "nixos-default-prefs.js")
+];
+```
+
+The code now clearly reads from top to bottom in the order the data is processed,
+it is easy to add and remove processing steps at any point.
+
+With a dedicated pipe operator, it would look like this:
+
+```nix
+defaultPrefsFile = defaultPrefs
+  |> lib.mapAttrsToList (
+    key: value: ''
+      // ${value.reason}
+      pref("${key}", ${builtins.toJSON value.value});
+    ''
+  )
+  |> lib.concatStringsSep "\n"
+  |> pkgs.writeText "nixos-default-prefs.js";
+```
+
+The artificial distinction between the first input and the functions via the list now is gone,
+and so are the parentheses around the functions.
+With the lower character overhead, using the operator becomes attractive in more situations,
+whereas a `pipe` pays for its overhead only in more complex scenarios (usually three functions or more).
+Having a dedicated operator also increases visibility and discoverability of the feature.
+
+# Detailed design
+[design]: #detailed-design
+
+## `|>` operator
+
+A new operator `|>` is introduced into the Nix language.
+It is defined as function application with the order of arguments swapped: `f a` = `a |> f`.
+It is left-associative and has a binding strength weaker than function application:
+`a |> f |> g b |> h` = `h ((g b) (f a))`.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## Tooling support
+
+Like any language extension, this will require the available Nix tooling to be updated.
+Updating parsers should be pretty easy, as the syntax changes to the language are fairly minimal.
+Tooling that evaluates Nix code in some way or does static code analysis should be easy to support too,
+since one may treat the operator as syntactic sugar for function application.
+No fundamentally new semantics are introduced to the language.
+
+## Nixpkgs interaction
+
+As soon as the Nixpkgs minimum version contains `|>`, using it will be allowed and encouraged in the documentation.
```

Comment:
Is it allowed to backport the pipe operator to the Nixpkgs minimum nix version?

Reply:
Do major feature additions like this usually get a backport?

Even if it was backported, wouldn't it defeat the purpose of having a minimal version that's likely to be installed by most users?

A backport would still be a new release, even if it was based on an old major version number.

Reply:
Sometimes.  Support for zstd cache compression was backported to 2.3.

The reason we have a minimum version of 2.3 not so much to support people who haven't updated Nix in several years, it's because later versions of Nix have [unresolved regressions](https://github.com/NixOS/nix/issues/9412) in things that some people depend on.